share/wake/lib/core/json.wake

```diff
@@ -538,3 +538,162 @@ export def lastValueInList (values: List JValue): Result JValue Error =
+    def view j =
+        require (JString s) = j else Nil
+        s, Nil
+    view
+
+export def jeString: (String => String) => JValue => JValue =
+    def edit f j =
+        require (JString s) = j else j
+        JString (f s)
+
+    edit
+
+export def jInteger: JValue => List Integer =
+    def view j =
+        require (JInteger ji) = j else Nil
+        ji, Nil
+
+    view
+
+export def jeInteger: (Integer => Integer) => JValue => JValue =
+    def edit f j =
+        require (JInteger ji) = j else j
+        JInteger (f ji)
+
+    edit
+
+export def jBoolean: JValue => List Boolean =
+    def view j =
+        require (JBoolean jb) = j else Nil
+        jb, Nil
+
+    view
+
+export def jeBoolean: (Boolean => Boolean) => JValue => JValue =
+    def edit f j =
+        require (JBoolean jb) = j else j
+        JBoolean (f jb)
+
+    edit
+
+# Are these safe since 1.0 => JDouble 1.0 => to_json => from_json => JInteger 1?
+# Maybe jNumber is the only *safe* thing to use? On the other hand I'm worried
+# about only giving jNumber as an abstraction over jDouble...IDK yet.
+export def jDouble: JValue => List Double =
+    def view j =
+        require (JDouble jd) = j else Nil
+        jd, Nil
+
+    view
+
+export def jeDouble: (Double => Double) => JValue => JValue =
+    def edit f j =
+        require (JDouble jd) = j else j
+        JDouble (f jd)
+
+    edit
+
+export def jNumber: JValue => List Double =
+    def view = match _
+        JDouble x  = x, Nil
+        JInteger x = fromOption $ dint x
+        _          = Nil
+
+    view
+
+# This is odd because if you edit a JInteger with say +1 you still get a JDouble
+export def jeNumber: (Double => Double) => JValue => JValue =
+    def edit f s = match s
+        JDouble x  = JDouble (f x)
+        JInteger x =
+            require Some dx = dint x else s
+            JDouble (f dx)
+        _          = s
+
+    edit
+
+export def jNull: JValue => List Unit =
+    def view j =
+        require JNull = j else Nil
+        Unit, Nil
+
+    view
+
+export def jeNull: (Unit => Unit) => JValue => JValue =
+    def edit f j =
+        require JNull = j else j
+        def _ = f Unit # the edit function might be effectful
+        JNull
+
+    edit
+
+export def jObject: JValue => List (List (Pair String JValue)) =
+    def view j =
+        require (JObject jo) = j else Nil
+        jo, Nil
+
+    view
+
+export def jeObject: (List (Pair String JValue) => List (Pair String JValue)) => JValue => JValue =
+    def edit f j =
+        require (JObject jo) = j else j
+        JObject (f jo)
+
+    edit
+
+export def jArray: JValue => List (List JValue) =
+    def view j =
+        require (JArray ja) = j else Nil
+        ja, Nil
+
+    view
+
+export def jeArray: (List JValue => List JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray jl) = j else j
+        JArray (f jl)
+
+    edit
+
+export def jSingelton: JValue => List JValue =
```

Comment:
Is there an actual use-case for this? Seems like a bitchy version of `jAt 0`.

Reply:
Yeah, its like the difference between `expectOne` and `expectFirst` lets you ensure that you're accessing exactly what you think you're accessing. Personally I think its generally best to write code that fails fast when any input diverges from the expectation; this is one way to do that.

I will say though that the error handling for this gets hidden so maybe this isn't good and in that case we should expect people to break things up into two queries. Like if you had `f |> jSingelton |> jField "foo"` and the result was empty you wouldn't know if it was because there were too many values in the list or if `foo` was missing so the error would have to be vauge or include multiple cases like `error: either ... was not a singleton or it was and the "foo" field was missing` which is a terrible error message.

I guess in light of that I'm inclined to remove it now but I do think we should keep the "expects a singleton" case in mind as a real case. It just just be handled with require and/or expectOne rather than this.

Reply:
I think I'd recommend using explicit `require` statements at the point where we check for the expected pattern, rather than an implicit delete-on-unexpected-input function.

Reply:
Yeah I agree now.